### ISSUE TITLE
No working console

### ISSUE THREAD

**Issue by drfrag666** [chocolate-doom/chocolate-doom#1211]
Could it be possible to easily add an enhanced console now with SDL2, one with the colored banner like DOS Doom and all? That would be pretty cool. I'm compiling my SDL1 based fork as both a GUI and console application. This is a serious issue specially for Crispy as you can't read warning messages.

**Comment by fabiangreffrath** [comment]
As I understand this issue, it depends on the underlying SDL2 library - i.e. if it has been compiled as a console library or not. For example, I usually build Crispy on Windows in an MSYS2 environment and this properly prints out all stdout/stderr messages to the console.

**Comment by drfrag666** [comment]
Really? For me it doesn't work with the Official Crispy build.  A new console is not open and even running it from the command prompt nothing gets printed there. There aren't even std* text files, not that it would be a solution. Could it be done with a GUI window or even text with colored output?

**Comment by JNechaevsky** [comment]
Nope, running from cmd.exe will not give a console output, some additional code must be implemented, but also probably declared as Windows-only. See https://github.com/chocolate-doom/chocolate-doom/issues/1166, I've made some research there, but have not finished it for Chocolate.

In short words: I've successfully implemented an optional console window, but was not lucky to fix "new line" issue while running from cmd.exe.

**Comment by fabiangreffrath** [comment]
Alright, the difference might be then that I am running the program from an actual terminal, i.e. mintty instead of plain cmd.exe.

**Comment by drfrag666** [comment]
Thanks! If you got colored banners that's great.
 I thought freopen was portable. I just did this and compiled the main exe with -mconsole. There's a side effect tough, running doom from setup to test controls enters an infinite loop unless you press ESC repeatedly.
```
--- a/src/doom/d_main.c
+++ b/src/doom/d_main.c
@@ -1223,6 +1223,9 @@ void D_DoomMain (void)
 
     I_AtExit(D_Endoom, false);
 
+    freopen("CONOUT$","w",stdout);
+    freopen("CONOUT$","w",stderr);
+
     // print banner
 
     I_PrintBanner(PACKAGE_STRING);

```

**Comment by JNechaevsky** [comment]
> If you got colored banners that's great.

By the way, is there portable existing approach for console coloring? For example, to make a colored title (red background, white text) on Windows, the code should be something like this, 80 chars long:

```
#ifdef _WIN32
#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif
#include <windows.h> // SetConsoleTextAttribute :(
#endif

...

{
SetConsoleTextAttribute(GetStdHandle(STD_OUTPUT_HANDLE), BACKGROUND_RED | FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE | FOREGROUND_INTENSITY);
    DEH_printf("                                Chocolate Doom " PACKAGE_VERSION
               "                                ");
    DEH_printf("\n");
}
```

But obviously, this will work only on Windows.

**Comment by drfrag666** [comment]
AFAIK you'd need to use a library for that. I guess you've noticed you don't need stdin, that's why you got the new line problem.
